z3c/dependencychecker/db.py

```diff
@@ -39,7 +44,234 @@ def add_extra_requirements(self, extra_name, dotted_names):
             self._extras_requirements[extra_name] = only_extra_dotted_names
 
     def _filter_duplicates(self, imports):
-        """Return all items in imports that are not found in reference"""
+        """Return all items in imports that are not a requirement already"""
         all_imports = set([dotted_name for dotted_name in imports])
         filtered = all_imports - self._requirements
         return filtered
+
+    def add_imports(self, imports):
+        for single_import in imports:
+            self.imports_used.append(single_import)
+
+    def get_missing_imports(self):
+        filters = (
+            self._filter_out_known_packages,
+            self._filter_out_testing_imports,
+            self._filter_out_own_package,
+            self._filter_out_requirements,
+            self._filter_out_python_standard_library,
+        )
+        missing = self._process_pipeline(self.imports_used, filters)
+        unique_imports = self._get_unique_imports(imports_list=missing)
+        return unique_imports
+
+    def get_missing_test_imports(self):
+        filters = (
+            self._filter_out_known_packages,
+            self._filter_out_only_testing_imports,
+            self._filter_out_own_package,
+            self._filter_out_requirements,
+            self._filter_out_test_requirements,
+            self._filter_out_python_standard_library,
+        )
+        missing = self._process_pipeline(self.imports_used, filters)
+        unique_imports = self._get_unique_imports(imports_list=missing)
+        return unique_imports
+
+    def get_unneeded_requirements(self):
+        all_but_test_requirements = self._requirements.copy()
+        for extra in self._extras_requirements:
+            all_but_test_requirements.update(self._extras_requirements[extra])
+        test_requirements = self._get_test_extra()
+        if test_requirements:
```

Comment:
Done! Actually that method ``._get_test_extra()`` was returning ``False`` if there was no extra, but I changed that to an empty list, so it will always return a list, which makes it easier.